[API] Return changeSet and validationMessages on dryrun response(s) t…

charlesg3 opened the pull request:
…o create/update cluster.

*Description of changes:*
- Add `validationMessages` to the response when creating a cluster if `dryrun` flag is set
- Add `validationMessages` and `changeSet` to the response when updating a cluster if `dryrun` flag is set
- Update unit tests to reflect above change
- Update integration tests to check for appropriate responses during create / update when doing dryrun

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

demartinofra commented on cli/tests/pcluster/api/controllers/test_cluster_operations_controller.py:
```diff
@@ -1207,10 +1207,22 @@ def test_successful_update_request(
     def test_dryrun(self, mocker, client):
```
nit: shall we also test the suppressed validation messages? Aren't we testing dryrun for create?

charlesg3 replied:
https://github.com/aws/aws-parallelcluster/pull/3122/commits/61fdaeae8c93895f9f0d9949604ac54ecb292d64

demartinofra reviewed (commented):
LGTM. Probably need to review integ tests + generate client code

charlesg3 commented:
> LGTM. Probably need to review integ tests + generate client code
Generate client code: https://github.com/aws/aws-parallelcluster/pull/3122/commits/a34fe87261a2ad9fbb128859401a1e526b4a8233

charlesg3 commented:
> LGTM. Probably need to review integ tests + generate client code

Updated integration tests: https://github.com/aws/aws-parallelcluster/pull/3122/commits/fe8b2b191e8549b116c7de1101c10b384f3d11ec
- Refactored the code to be more data-driven -- describes the various combinations of suppression / warning vs error / output / dryrun (and it is validated for both create and update operations.)

demartinofra commented on tests/integration-tests/tests/cli_commands/test_cli_commands.py:
```diff
@@ -357,61 +303,59 @@ def _test_pcluster_export_cluster_logs(s3_bucket_factory, cluster):
     logging.info("Testing that pcluster list-cluster-log-streams is working as expected")
     list_streams_result = cluster.list_log_streams()
+    cluster_info = cluster.describe_cluster()
     streams = list_streams_result["items"]
```
probably this needs to be updated

demartinofra commented on tests/integration-tests/utils.py:
```diff
@@ -208,12 +204,9 @@ def get_cluster_nodes_instance_ids(stack_name, region, instance_types=None, node
+        return [instance["InstanceId"] for instance in instances]
     except Exception as e:
-        logging.error("Failed retrieving instance ids with exception: {}".format(e))
+        logging.error("Failed retrieving instance ids with exception: %", e)
```
%s?
